ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Byron/Aux.hs

```diff
@@ -0,0 +1,545 @@
+    CBOR.decodeWord8 >>= \case
+      0   -> MempoolTxErr             <$> fromCBOR
+      1   -> MempoolDlgErr            <$> fromCBOR
+      2   -> MempoolUpdateProposalErr <$> fromCBOR
+      3   -> MempoolUpdateVoteErr     <$> fromCBOR
+      tag -> cborError $ DecoderErrorUnknownTag "ApplyMempoolPayloadErr" tag
+
+applyMempoolPayload :: MonadError ApplyMempoolPayloadErr m
+                    => CC.ValidationMode
+                    -> Gen.Config
+                    -> CC.AMempoolPayload ByteString
+                    -> CC.ChainValidationState -> m CC.ChainValidationState
+applyMempoolPayload validationMode cfg payload =
+    case payload of
+      CC.MempoolTx tx ->
+        (`wrapError` MempoolTxErr) .
+          applyTxAux validationMode cfg tx
+      CC.MempoolDlg cert ->
+        (`wrapError` MempoolDlgErr) .
+          applyCertificate cfg cert
+      CC.MempoolUpdateProposal proposal ->
+        (`wrapError` MempoolUpdateProposalErr) .
+          applyUpdateProposal cfg proposal
+      CC.MempoolUpdateVote vote ->
+        (`wrapError` MempoolUpdateVoteErr) .
+          applyUpdateVote cfg vote
+
+-- | The encoding of the mempool payload (without a 'AMempoolPayload' envelope)
+mempoolPayloadRecoverBytes :: CC.AMempoolPayload ByteString -> ByteString
+mempoolPayloadRecoverBytes = go
+  where
+    go :: CC.AMempoolPayload ByteString -> ByteString
+    go (CC.MempoolTx             payload) = recoverBytes payload
+    go (CC.MempoolDlg            payload) = recoverBytes payload
+    go (CC.MempoolUpdateProposal payload) = recoverBytes payload
+    go (CC.MempoolUpdateVote     payload) = recoverBytes payload
+
+-- | Re-encode the mempool payload (without any envelope)
+mempoolPayloadReencode :: CC.AMempoolPayload a -> ByteString
+mempoolPayloadReencode = go
+  where
+    go (CC.MempoolTx             payload) = reencode payload
+    go (CC.MempoolDlg            payload) = reencode payload
+    go (CC.MempoolUpdateProposal payload) = reencode payload
+    go (CC.MempoolUpdateVote     payload) = reencode payload
+
+    reencode :: (Functor f, ToCBOR (f ())) => f a -> ByteString
+    reencode = CBOR.toStrictByteString . toCBOR . void
+
+{-------------------------------------------------------------------------------
+  Annotations
+-------------------------------------------------------------------------------}
+
+reAnnotateMagicId :: ProtocolMagicId -> Annotated ProtocolMagicId ByteString
+reAnnotateMagicId pmi = reAnnotate $ Annotated pmi ()
+
+reAnnotateMagic :: ProtocolMagic -> AProtocolMagic ByteString
+reAnnotateMagic (AProtocolMagic a b) = AProtocolMagic (reAnnotate a) b
+
+reAnnotateBlock :: CC.EpochSlots -> CC.ABlock () -> CC.ABlock ByteString
+reAnnotateBlock epochSlots =
+      (\bs -> splice bs $ CBOR.deserialiseFromBytes
+                            (CC.fromCBORABlock epochSlots)
+                            bs)
+    . CBOR.toLazyByteString
+    . CC.toCBORBlock epochSlots
+  where
+    splice :: Show err
```

Comment:
Generalize `splice`, then use it in tests (`Test.Consensus.Ledger.Byron`, see `annotate`).

Reply:
Now available as `reAnnotateUsing`.